### PR TITLE
CodeWhisperer security scans support for Ruby files. 

### PR DESCRIPTION
**Pull request by laileni-aws**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [x] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by ctlai95 — APPROVED** [review]

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererRubyCodeScanTest.kt [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+            onGeneric { getCodeScan(any(), any()) }.thenReturn(fakeGetCodeScanResponse)
+            onGeneric { listCodeScanFindings(any(), any()) }.thenReturn(fakeListCodeScanFindingsResponse)
+        }
+    }
+
+    @Test
+    fun `test createPayload`() {
+        val payload = sessionConfigSpy.createPayload()
+        assertNotNull(payload)
+        assertThat(payload.context.totalFiles).isEqualTo(3)
```
Comment (Will-ShaoHua) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererRubyCodeScanTest.kt:52]: same here, better to assert the contents of the list

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererRubyCodeScanTest.kt [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    }
+
+    @Test
+    fun `test getImportedFiles()`() {
+        val files = sessionConfigSpy.getImportedFiles(testRuby, setOf())
+        assertNotNull(files)
+        assertThat(files).hasSize(2)
+        assertThat(files).contains(utilsRuby.path)
+        assertThat(files).contains(helperRuby.path)
+    }
```
[the remark this answers is not in the document]
Comment (Will-ShaoHua) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererRubyCodeScanTest.kt:96]: better to cover more cases as the regex intends 
```
Regex("^(require|require_relative|load|include|extend)\\s+('[^']+'|\"[^\"]+\"|\\w+)(\\s+as\\s+(\\w+))?")
```

**Comment by Will-ShaoHua** [comment]
not a blocker to this PR since for all our previous implementation we have been relying on regex to manually parse import statements, but wondering have we tried use `psiFile.importList?.allImportStatements` to get imports? Ideally IDE should do the heavy lift work for us, we can still provide a backup plan(regex) IF the IDE doesn't recognize the file (for example opening a TS/JS file in JB community). 

here is an example
https://github.com/aws/aws-toolkit-jetbrains/blob/main/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/JavaCodeWhispererFileCrawler.kt#L33

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererConstants.kt [hunk 3]
```diff
@@ -39,6 +39,12 @@ object CodeWhispererConstants {
     const val USING: String = "using"
     const val GLOBAL_USING: String = "global using"
     const val STATIC: String = "static"
+    const val REQUIRE: String = "require"
+    const val REQUIRE_RELATIVE: String = "require_relative"
+    const val LOAD: String = "load"
+    const val INCLUDE: String = "include"
+    const val EXTEND: String = "extend"
+    const val AS: String = " as "
```
Comment (Will-ShaoHua) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererConstants.kt:47]: nit: I feel this list will keep growing, maybe we can consider move these syntax keywords to a different constant file

Comment (laileni-aws) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererConstants.kt:47]: Yes, I agree, I am working on this, to raise a PR after this PR, with optimization in this constants file, implementation and testing files of security scans as their is some duplicate code.

Made a note and I will move these constant strings to separate file.

**Review by Will-ShaoHua — APPROVED** [review]

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererRubyCodeScanTest.kt (hunk 1 above)
Comment (laileni-aws) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererRubyCodeScanTest.kt]: Payload does have a parameter like number of files present but does not have exact file names. So I am not able to check the fileNames for all files present but we can do this for scanned files though. Correct me If I am wrong @Will-ShaoHua 
Thank you


image

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererRubyCodeScanTest.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (laileni-aws) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererRubyCodeScanTest.kt:101]: Added these use cases

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererRubyCodeScanTest.kt (hunk 1 above)
Comment (Will-ShaoHua) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererRubyCodeScanTest.kt]: oh nm, I thought it's a list but it's not

Comment (Will-ShaoHua) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererRubyCodeScanTest.kt]: thanks for the clarification

**Comment by laileni-aws** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=aws_aws-toolkit-jetbrains&pullRequest=4033)
> 
> [Bug](https://sonarcloud.io/project/issues?id=aws_aws-toolkit-jetbrains&pullRequest=4033&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=aws_aws-toolkit-jetbrains&pullRequest=4033&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=aws_aws-toolkit-jetbrains&pullRequest=4033&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=aws_aws-toolkit-jetbrains&pullRequest=4033&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=aws_aws-toolkit-jetbrains&pullRequest=4033&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=aws_aws-toolkit-jetbrains&pullRequest=4033&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=aws_aws-toolkit-jetbrains&pullRequest=4033&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=aws_aws-toolkit-jetbrains&pullRequest=4033&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=aws_aws-toolkit-jetbrains&pullRequest=4033&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=aws_aws-toolkit-jetbrains&pullRequest=4033&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=aws_aws-toolkit-jetbrains&pullRequest=4033&resolved=false&types=CODE_SMELL) [9 Code Smells](https://sonarcloud.io/project/issues?id=aws_aws-toolkit-jetbrains&pullRequest=4033&resolved=false&types=CODE_SMELL)
> 
> [No Coverage information](https://sonarcloud.io/component_measures?id=aws_aws-toolkit-jetbrains&pullRequest=4033) No Coverage information [](https://sonarcloud.io/component_measures?id=aws_aws-toolkit-jetbrains&pullRequest=4033&metric=new_duplicated_lines_density&view=list) [6.9% Duplication](https://sonarcloud.io/component_measures?id=aws_aws-toolkit-jetbrains&pullRequest=4033&metric=new_duplicated_lines_density&view=list)
> 
> idea Catch issues before they fail your Quality Gate with our IDE extension sonarlint [SonarLint](https://www.sonarsource.com/products/sonarlint/features/connected-mode/?referrer=sonarcloud-welcome)

I will raise a follow up CR to reduce the code duplication not only ruby but for all security scans which effects all the security scans test files. So, In this PR, I am focusing wrt to Ruby.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/sessionconfig/RubyCodeScanSessionConfig.kt [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        for (i in inputPaths.indices) {
+            val outputPath = inputPaths.subList(0, i + 1).joinToString(FILE_SEPARATOR.toString())
+            outputPaths.add(outputPath)
+        }
+        return outputPaths
+    }
+
+    private fun getModulePath(modulePathStr: String): MutableSet<String> {
+        val pos = modulePathStr.indexOf(" ${CodeWhispererConstants.AS} ")
+        var modifiedModulePathStr = modulePathStr
```
[the remark this answers is not in the document]
Comment (laileni-aws) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/sessionconfig/RubyCodeScanSessionConfig.kt]: Updated Accordingly.

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererCodeScanTestBase.kt [hunk 5]
```diff
@@ -226,6 +229,39 @@ open class CodeWhispererCodeScanTestBase(projectRule: CodeInsightTestFixtureRule
...
         ]                
     """
+    internal fun getSourceFilesUnderProjectRoot(sessionConfigSpy: CodeScanSessionConfig, testFile: VirtualFile, size: Int) {
+        assertThat(sessionConfigSpy.getSourceFilesUnderProjectRoot(testFile).size).isEqualTo(size)
+    }
+    internal fun getTotalProjectSizeInBytes(sessionConfigSpy: CodeScanSessionConfig, totalSize: Long) {
+        runBlocking {
+            assertThat(sessionConfigSpy.getTotalProjectSizeInBytes()).isEqualTo(totalSize)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (laileni-aws) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/codescan/CodeWhispererCodeScanTestBase.kt]: Apologies! This file is not required for this PR!
Updated Accordingly.

**Review by manodnyab — APPROVED** [review]

**Pull request merged**